node/index.js

```diff
@@ -867,4 +759,259 @@ function TChannelClientOp(req, start) {
+        sock.destroy();
+    });
+    function onClose() {
+        if (--counter <= 0) {
+            if (counter < 0) {
+                self.logger.error('closed more sockets than expected', {
+                    counter: counter
+                });
+            }
+            callback();
+        }
+    }
+};
+
+TChannelPeer.prototype.setState = function setState(state) {
+    var self = this;
+    var oldState = self.state;
+    self.state = state;
+    self.emit('stateChanged', oldState, state);
+};
+
+TChannelPeer.prototype.connect = function connect() {
+    var self = this;
+    var conn = self.connections[self.connections.length - 1];
+    if (!conn) {
+        var socket = self.makeOutSocket();
+        conn = self.makeOutConnection(socket);
+        self.addConnection(conn);
+    }
+    return conn;
+};
+
+TChannelPeer.prototype.request = function request(options) {
+    var self = this;
+    return self.connect().request(options);
+};
+
+TChannelPeer.prototype.addConnection = function addConnection(conn) {
+    var self = this;
+    // TODO: first approx alert for self.connections.length > 2
+    // TODO: second approx support pruning
+    if (conn.direction === 'out') {
+        self.connections.push(conn);
+    } else {
+        self.connections.unshift(conn);
+    }
+    return conn;
+};
+
+TChannelPeer.prototype.removeConnection = function removeConnection(conn) {
+    var self = this;
+    var list = self.connections;
+    var index = list ? list.indexOf(conn) : -1;
+    if (index !== -1) {
+        return list.splice(index, 1)[0];
+    } else {
+        return null;
+    }
+};
+
+TChannelPeer.prototype.makeOutSocket = function makeOutSocket() {
+    var self = this;
+    var parts = self.hostPort.split(':');
+    if (parts.length !== 2) {
+        throw new Error('invalid destination'); // TODO typed error
+    }
+    var host = parts[0];
+    var port = parts[1];
+    if (host === '0.0.0.0' || port === '0') {
+        throw new Error('cannot make out connection to ephemeral peer'); // TODO typed error
+    }
+    var socket = net.createConnection({host: host, port: port});
+    return socket;
+};
+
+
+TChannelPeer.prototype.makeOutConnection = function makeOutConnection(socket) {
+    var self = this;
+    var conn = new TChannelConnection(self.channel, socket, 'out', self.hostPort);
+    conn.once('reset', onConnectionReset);
+    self.emit('allocConnection', conn);
+    return conn;
+
+    function onConnectionReset(/* err */) {
+        // TODO: log?
+        self.removeConnection(conn);
+    }
+};
+
+function TChannelPeerState(channel, name) {
+    var self = this;
+    self.channel = channel;
+    self.name = name;
+}
+
+TChannelPeerState.prototype.shouldRequest = function shouldRequest(/* options, op */) {
+    // TODO: op isn't quite right currently as a "TChannelClientOp", the
+    // intention is that the other (non-options) arg encapsulates all requests
+    // across retries and setries
+    return 0;
```

Comment:
What does `0` means ? I would guess that `0` means that it should never request.